src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,335 @@
+            );
+        }
+
+        $this->mongo = $mongo;
+
+        $this->options = array_merge(array(
+            'collection' => 'lock',
+        ), $options);
+
+        $this->initialTtl = $initialTtl;
+    }
+
+    /**
+     * Create a TTL index to automatically remove expired locks.
+     *
+     * This should be called once during database setup.
+     *
+     * Alternatively the TTL index can be created manually:
+     *
+     *  db.lock.ensureIndex(
+     *      { "expires_at": 1 },
+     *      { "expireAfterSeconds": 0 }
+     *  )
+     *
+     * Please note, expires_at is based on the application server. If the
+     * database time differs; a lock could be cleaned up before it has expired.
+     * Set a positive expireAfterSeconds to account for any time drift between
+     * application and database server.
+     *
+     * A TTL index MUST BE used on MongoDB 2.2+ to automatically clean up expired locks.
+     *
+     * @see http://docs.mongodb.org/manual/tutorial/expire-data/
+     *
+     * @param int $expireAfterSeconds Number of seconds after expiry to wait before deleting
+     *
+     * @return string The name of the created index
+     *
+     * @throws UnsupportedException          if options are not supported by the selected server
+     * @throws MongoInvalidArgumentException for parameter/option parsing errors
+     * @throws DriverRuntimeException        for other driver errors (e.g. connection errors)
+     */
+    public function createTtlIndex(int $expireAfterSeconds = 0): string
+    {
+        $keys = array(
+            'expires_at' => 1,
+        );
+
+        $options = array(
+            'expireAfterSeconds' => $expireAfterSeconds,
+        );
+
+        return $this->getCollection()->createIndex($keys, $options);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws LockExpiredException when save is called on an expired lock
+     */
+    public function save(Key $key)
+    {
+        $token = $this->getUniqueToken($key);
+        $now = microtime(true);
+
+        $filter = array(
+            '_id' => (string) $key,
+            '$or' => array(
+                array(
+                    'token' => $token,
+                ),
+                array(
+                    'expires_at' => array(
+                        '$lte' => $this->createDateTime($now),
+                    ),
+                ),
+            ),
+        );
+
+        $update = array(
+            '$set' => array(
+                '_id' => (string) $key,
+                'token' => $token,
+                'expires_at' => $this->createDateTime($now + $this->initialTtl),
+            ),
+        );
+
+        $options = array(
+            'upsert' => true,
+        );
+
+        $key->reduceLifetime($this->initialTtl);
+
+        try {
+            $this->getCollection()->updateOne($filter, $update, $options);
+        } catch (WriteException $e) {
+            $writeErrors = $e->getWriteResult()->getWriteErrors();
```

Comment:
You could extract this logic into a function to follow DRY principle :)

Reply:
moved